stmhal/sdcard.c

```diff
@@ -85,6 +86,29 @@ void sdcard_init(void) {
     // invalidate the sd_handle
     sd_handle.Instance = NULL;
 
+    // configure the SD card detect pin
+    // we do this here so we can detect if the SD card is inserted before powering it on
+    GPIO_Init_Structure.Mode = GPIO_MODE_INPUT;
+    GPIO_Init_Structure.Speed = GPIO_SPEED_LOW;
+#if defined(MICROPY_HW_SDCARD_DETECT_PIN)
+    // assume a mechanical switch is used
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = MICROPY_HW_SDCARD_DETECT_PIN.pin_mask;
+    HAL_GPIO_Init(MICROPY_HW_SDCARD_DETECT_PIN.gpio, &GPIO_Init_Structure);
+#elif defined(MICROPY_HW_SDCARD_DETECT_PULL)
+    // assume no mechanical switch, so assume CD/DAT3 is used for card detect.
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = pin_C11.pin_mask;
+    HAL_GPIO_Init(pin_C11.gpio, &GPIO_Init_Structure);
+    // capture sdcard present state before the IO pin is remapped to SDIO functionality
+    sdcard_present = (HAL_GPIO_ReadPin(pin_C11.gpio, pin_C11.pin_mask) == MICROPY_HW_SDCARD_DETECT_PRESENT);
+    // TODO: do we need to issue ACMD42 (SD_CMD_SD_APP_SET_CLR_CARD_DETECT) to disable the internal pullup resistor?
+#else
+    // SD Card Detect not defined.  Just assume it's there.
```

Comment:
Ok.  Does teensy have any way to detect the sdcard?  If not I'd say that's a big oversight.  I don't want to have to implement/support many ways of detecting the SD card, especially if each board does their own thing.